### PR TITLE
Lora scorer

### PR DESCRIPTION
**Pull request by riapush**

### CODE REVIEW

File: autointent/modules/scoring/_lora/lora.py [hunk 1]
```diff
@@ -0,0 +1,198 @@
...
+        self._model = AutoModelForSequenceClassification.from_pretrained(model_name, num_labels=num_labels)
+
+        # Configure LoRA
+        lora_config = LoraConfig(
+            r=self.lora_rank,  # Rank of the low-rank matrices
+            lora_alpha=self.lora_alpha,  # Scaling factor
+            target_modules=["query", "value"],  # Target modules to apply LoRA
+            lora_dropout=self.lora_dropout,  # Dropout rate for LoRA layers
+            bias="none",  # Whether to add bias to LoRA layers
+        )
```
Comment (voorhs) [inline autointent/modules/scoring/_lora/lora.py]: предлагаю не менять точечно параметры `LoraConfig`, а вместо этого сделать такой конструктор:
```
def __init__(..., **lora_kwargs: Any):
    ...
    self._lora_config = LoraConfig(**lora_kwargs)
    ...
```

File: autointent/modules/scoring/_lora/lora.py
Comment (voorhs) [inline autointent/modules/scoring/_lora/lora.py:1]: надо будет сильно зарефакторить после мерджа #165

**Review by voorhs — CHANGES_REQUESTED** [review]

File: autointent/modules/scoring/_lora/lora.py [hunk 2]
```diff
@@ -0,0 +1,160 @@
...
+    TrainingArguments,
+)
+
+from autointent import Context
+from autointent.configs import HFModelConfig
+from autointent.custom_types import ListOfLabels
+from autointent.modules.base import BaseScorer
+
+
+class BERTLoRAScorer(BaseScorer):
```
Comment (voorhs) [inline autointent/modules/scoring/_lora/lora.py:25]: надо наследоваться от `autointent.modules.base.ScoringModule`

в #165 когда делал ревью этого не заметил

надо бы в #174 или #172 исправить это

File: autointent/modules/scoring/_lora/lora.py [hunk 3]
```diff
@@ -0,0 +1,160 @@
...
+    name = "lora"
+    supports_multiclass = True
+    supports_multilabel = True
+    _multilabel: bool
```
Comment (voorhs) [inline autointent/modules/scoring/_lora/lora.py]: эту строчку можно убрать, она прописана в базовом классе

File: pyproject.toml [hunk 4]
```diff
@@ -45,6 +45,7 @@ dependencies = [
     "xxhash (>=3.5.0,<4.0.0)",
     "python-dotenv (>=1.0.1,<2.0.0)",
     "transformers[torch] (>=4.49.0,<5.0.0)",
+    "peft (==0.10.0)",
```
Comment (voorhs) [inline pyproject.toml]: надо убрать ограничение == и сделать  хотя бы `>= 0.10.0, <1.0.0`

на этой старой версии mypy выдает ошибки, которые были исправлены недавно: https://github.com/huggingface/peft/pull/2265

File: autointent/modules/scoring/_lora/lora.py [hunk 5]
```diff
@@ -0,0 +1,160 @@
...
+            **lora_kwargs,
+        )
+
+    def get_embedder_config(self) -> dict[str, Any]:
+        return self.model_config.model_dump()
+
+    def _validate_task(self, labels: ListOfLabels) -> None:
+        """Validate the task and set _multilabel flag."""
+        super()._validate_task(labels)
+        self._multilabel = isinstance(labels[0], list)
```
Comment (voorhs) [inline autointent/modules/scoring/_lora/lora.py]: этот метод нет необходимости переопределять, а проверка на multilabel уже реализована в базовом классе

File: autointent/modules/scoring/_lora/lora.py [hunk 6]
```diff
@@ -0,0 +1,160 @@
...
+        if hasattr(self, "_model"):
+            self.clear_cache()
+
+        self._validate_task(labels)
+
+        if self._multilabel:
+            labels_array = np.array(labels)
+            num_labels = labels_array.shape[1]
+        else:
+            num_labels = len(set(labels))
```
Comment (voorhs) [inline autointent/modules/scoring/_lora/lora.py]: это тоже не нужно, потому что `self._validate_task(labels)` сохраняет число классов в `self._n_classes`

File: autointent/modules/scoring/_lora/lora.py [hunk 7]
```diff
@@ -0,0 +1,160 @@
...
+
+        model_name = self.model_config.model_name
+        self._tokenizer = AutoTokenizer.from_pretrained(model_name)
+        self._model = AutoModelForSequenceClassification.from_pretrained(model_name, num_labels=num_labels)
+
+        # Apply LoRA to the model
+        self._model = get_peft_model(self._model, self._lora_config)
+
+        use_cpu = hasattr(self.model_config, "device") and self.model_config.device == "cpu"
```
Comment (voorhs) [inline autointent/modules/scoring/_lora/lora.py]: model_config всегда имеет атрибут device, нет необходимости в этой проверке

File: autointent/modules/scoring/_lora/lora.py [hunk 8]
```diff
@@ -0,0 +1,160 @@
...
+            training_args = TrainingArguments(
+                output_dir=tmp_dir,
+                num_train_epochs=self.num_train_epochs,
+                per_device_train_batch_size=self.batch_size,
+                learning_rate=self.learning_rate,
+                seed=self.seed,
+                save_strategy="no",
+                logging_strategy="steps",
+                logging_steps=10,
+                report_to="wandb",
```
Comment (voorhs) [inline autointent/modules/scoring/_lora/lora.py]: аргумент `report_to` надо доставать из контексте при иницилиализации в `from_context`

вот референс: https://github.com/deeppavlov/AutoIntent/blob/644a849a336106edead81b400ed7b30e14a03fe6/autointent/modules/scoring/_bert.py#L61

File: autointent/modules/scoring/_lora/lora.py [hunk 9]
```diff
@@ -0,0 +1,160 @@
...
+        self._model.eval()
+
+    def predict(self, utterances: list[str]) -> npt.NDArray[Any]:
+        if not hasattr(self, "_model") or not hasattr(self, "_tokenizer"):
+            msg = "Model is not trained. Call fit() first."
+            raise RuntimeError(msg)
+
+        inputs = self._tokenizer(
+            utterances, padding=True, truncation=True, max_length=self.model_config.tokenizer_config.max_length, return_tensors="pt"
+        )
```
Comment (voorhs) [inline autointent/modules/scoring/_lora/lora.py]: аргументы токенизатора надо доставать из tokenizer_config

File: autointent/modules/scoring/_lora/lora.py [hunk 10]
```diff
@@ -0,0 +1,160 @@
...
+
+        with torch.no_grad():
+            outputs = self._model(**inputs)
+            logits = outputs.logits
```
Comment (voorhs) [inline autointent/modules/scoring/_lora/lora.py]: надо добавить батчированный инференс и перенос на девайс

вот референс https://github.com/deeppavlov/AutoIntent/blob/644a849a336106edead81b400ed7b30e14a03fe6/autointent/modules/scoring/_bert.py#L125

**Review by voorhs — CHANGES_REQUESTED** [review]

File: autointent/modules/scoring/_lora/lora.py [hunk 11]
```diff
@@ -0,0 +1,151 @@
...
+    _tokenizer: Any
+
+    def __init__(
+        self,
+        model_config: HFModelConfig | str | dict[str, Any] | None = None,
+        num_train_epochs: int = 3,
+        batch_size: int = 8,
+        learning_rate: float = 5e-5,
+        seed: int = 0,
+        report_to: REPORTERS_NAMES | None = None,  # type: ignore  # noqa: PGH003
```
Comment (voorhs) [inline autointent/modules/scoring/_lora/lora.py]: лучше указывать конкретный тип mypy ошибки которую ты игнорируешь, например `# type: ignore[no-any-return]`

File: autointent/modules/scoring/_lora/lora.py [hunk 12]
```diff
@@ -0,0 +1,151 @@
...
+        self.seed = seed
+        self.report_to = report_to
+        self._lora_config = LoraConfig(**lora_kwargs)
+
+    @classmethod
+    def from_context(
+        cls,
+        context: Context,
+        model_config: HFModelConfig | str | dict[str, Any] | None = None,
+        num_train_epochs: int = 10,
```
Comment (voorhs) [inline autointent/modules/scoring/_lora/lora.py]: тут надо поставить дефолтное значение 3 эпохи, как в конструкторе

чисто чтобы расхождений не было

File: autointent/modules/scoring/_lora/lora.py [hunk 13]
```diff
@@ -0,0 +1,151 @@
...
+            inputs = self._tokenizer(batch, return_tensors="pt", **self.model_config.tokenizer_config.model_dump())
+            with torch.no_grad():
+                outputs = self._model(**inputs)
+                logits = outputs.logits
+            if self._multilabel:
+                batch_predictions = torch.sigmoid(logits).numpy()
+            else:
+                batch_predictions = torch.softmax(logits, dim=1).numpy()
+            all_predictions.append(batch_predictions)
+        return np.vstack(all_predictions) if all_predictions else np.array([])
```
Comment (voorhs) [inline autointent/modules/scoring/_lora/lora.py]: нет переноса тензоров на девайс

File: autointent/modules/scoring/_lora/lora.py [hunk 14]
```diff
@@ -0,0 +1,151 @@
...
+        labels: ListOfLabels,
+    ) -> None:
+        if hasattr(self, "_model"):
+            self.clear_cache()
+
+        self._validate_task(labels)
+
+        model_name = self.model_config.model_name
+        self._tokenizer = AutoTokenizer.from_pretrained(model_name)
+        self._model = AutoModelForSequenceClassification.from_pretrained(model_name, num_labels=self._n_classes)
```
Comment (voorhs) [inline autointent/modules/scoring/_lora/lora.py]: надо еще передавать параметр `problem_type` который принимает значения либо `single_label_classification` либо `multi_label_classification`

File: autointent/modules/scoring/_lora/lora.py [hunk 15]
```diff
@@ -0,0 +1,151 @@
...
+        if model_config is None:
+            model_config = context.resolve_embedder()
+        return cls(
+            model_config=model_config,
+            num_train_epochs=num_train_epochs,
+            batch_size=batch_size,
+            learning_rate=learning_rate,
+            seed=seed,
+            report_to=context.logging_config.report_to
+            **lora_kwargs,
```
Comment (voorhs) [inline autointent/modules/scoring/_lora/lora.py]: ты запятую забыла в 70 строчке, из-за этого ошибка mypy

**Review by voorhs — CHANGES_REQUESTED** [review]

**Comment by voorhs** [comment]
на инференсе у тебя падает потому что метки для мультилейбла надо сконвертить во float перед подачей в Trainer

вот референс: https://github.com/deeppavlov/AutoIntent/blob/feat/ptuning_scorer/autointent/modules/scoring/_ptuning/ptuning.py#L152

**Comment by voorhs** [comment]
а вообще, как я и сказал надо двигаться к тому чтобы наследоваться от bert scorer

File: autointent/modules/scoring/_lora/lora.py [hunk 16]
```diff
@@ -0,0 +1,167 @@
...
+
+    def __init__(
+        self,
+        model_config: HFModelConfig | str | dict[str, Any] | None = None,
+        num_train_epochs: int = 3,
+        batch_size: int = 8,
+        learning_rate: float = 5e-5,
+        seed: int = 0,
+        report_to: REPORTERS_NAMES | None = None,  # type: ignore[no-any-return]
+        **lora_kwargs: Any, # noqa: ANN401
```
Comment (Samoed) [inline autointent/modules/scoring/_lora/lora.py]:
```suggestion
        **lora_kwargs: dict[str, Any],
```

File: autointent/modules/scoring/_lora/lora.py [hunk 17]
```diff
@@ -0,0 +1,167 @@
...
+    @classmethod
+    def from_context(
+        cls,
+        context: Context,
+        model_config: HFModelConfig | str | dict[str, Any] | None = None,
+        num_train_epochs: int = 3,
+        batch_size: int = 8,
+        learning_rate: float = 5e-5,
+        seed: int = 0,
+        **lora_kwargs: Any, # noqa: ANN401
```
Comment (Samoed) [inline autointent/modules/scoring/_lora/lora.py]:
```suggestion
        **lora_kwargs: dict[str, Any],
```

File: autointent/modules/scoring/_lora/lora.py [hunk 18]
```diff
@@ -0,0 +1,167 @@
...
+class BERTLoRAScorer(BaseScorer):
+    name = "lora"
+    supports_multiclass = True
+    supports_multilabel = True
+    _model: Any
+    _tokenizer: Any
+
+    def __init__(
+        self,
+        model_config: HFModelConfig | str | dict[str, Any] | None = None,
```
Comment (Samoed) [inline autointent/modules/scoring/_lora/lora.py]: Поменяй название model_config на что-то другое

Comment (voorhs) [reply autointent/modules/scoring/_lora/lora.py]: например `classification_model_config` или `transformer_config`

File: tests/modules/scoring/test_lora.py
Comment (voorhs) [inline tests/modules/scoring/test_lora.py:1]: Надо в тест добавить в конце проверку что после dump load предикшены не изменились

Comment (voorhs) [inline tests/modules/scoring/test_lora.py:1]: Надо в тест добавить в конце проверку что после dump load предикшены не изменились

File: autointent/modules/scoring/_lora/lora.py [hunk 19]
```diff
@@ -0,0 +1,167 @@
...
+
+        self._validate_task(labels)
+
+        model_name = self.model_config.model_name
+        self._tokenizer = AutoTokenizer.from_pretrained(model_name)
+        self._model = AutoModelForSequenceClassification.from_pretrained(
+            model_name,
+            num_labels=self._n_classes,
+            problem_type="multi_label_classification" if self._multilabel else "single_label_classification"
+            )
```
Comment (Samoed) [inline autointent/modules/scoring/_lora/lora.py]: В https://github.com/deeppavlov/AutoIntent/pull/185 я добавил `trust_remote_code` в конфиг моделей. Сюда тоже надо добавить

File: autointent/modules/scoring/_lora/lora.py [hunk 20]
```diff
@@ -0,0 +1,71 @@
...
+from autointent.configs import HFModelConfig
+from autointent.modules.scoring._bert import BertScorer
+
+
+class BERTLoRAScorer(BertScorer):
+    name = "lora"
+    supports_multiclass = True
+    supports_multilabel = True
+    _model: Any
+    _tokenizer: Any
```
Comment (voorhs) [inline autointent/modules/scoring/_lora/lora.py]: добавь докстринг в котором будут
- аргументы конструктора
- ссылка на документацию пефта с возможными параметрами лоры 
- пример использования твоего модуля

за образец можешь взять докстрингу у никиты: https://github.com/deeppavlov/AutoIntent/blob/feat/ptuning_scorer/autointent/modules/scoring/_ptuning/ptuning.py#L26

File: tests/assets/configs/multiclass.yaml [hunk 21]
```diff
@@ -35,6 +35,13 @@
       batch_size: [8, 16]
       learning_rate: [5.0e-5]
       seed: [0]
+    - module_name: lora
+      classification_model_config:
+        - model_name: avsolatorio/GIST-small-Embedding-v0
+      num_train_epochs: [1]
+      batch_size: [8, 16]
+      learning_rate: [5.0e-5]
+      seed: [0]
```
Comment (voorhs) [inline tests/assets/configs/multiclass.yaml:44]: добавь пожалуйста сюда еще какой-нибудь один параметр лоры

File: tests/assets/configs/multilabel.yaml [hunk 22]
```diff
@@ -31,6 +31,13 @@
       batch_size: [8]
       learning_rate: [5.0e-5]
       seed: [0]
+    - module_name: lora
+      classification_model_config:
+        - model_name: avsolatorio/GIST-small-Embedding-v0
+      num_train_epochs: [1]
+      batch_size: [8]
+      learning_rate: [5.0e-5]
+      seed: [0]
```
Comment (voorhs) [inline tests/assets/configs/multilabel.yaml:40]: и сюда

**Review by voorhs — CHANGES_REQUESTED** [review]
почти готово!

**Review by voorhs — APPROVED** [review]
ура ура ура
